### ISSUE TITLE
Fastify support

### ISSUE THREAD

**Issue by brodo** [SAP/cf-nodejs-logging-support#191]
It would be nice if this library would be compatible with [fastify](https://www.fastify.io).

**Comment by saard** [comment]
+1 fastify is widely used framework and it would be great if you can support it.

**Comment by christiand93** [comment]
Hi,
thank you for the proposal. I will check wether we can add fastify support asap.

**Comment by saard** [comment]
Hi @christiand93 Is there any outlook on supporting fastify?

**Comment by christiand93** [comment]
Hi @saard,
I have created a work-in-progress [PR](https://github.com/SAP/cf-nodejs-logging-support/pull/201) with a middleware implementation for Fastify. It needs some more polishing, tests etc. But the concept seems to work so far.
If you want to give a try (and share some feedback), you can check out the [branch](https://github.com/SAP/cf-nodejs-logging-support/tree/fastify) and build the library using `npm i` and `npm run build`. Setup a hook for the logging library as follows:

```js
var log = require('path/to/cf-nodejs-logging-support/build/main') // <-- set correct path
log.setFramework(log.Framework.Fastify)

const fastify = require('fastify')()

fastify.addHook("onRequest", log.logNetwork)

fastify.get('/', function (request, reply) {
    request.logger.info("message logged w/ request context")
    reply.send({ hello: 'world' })
})

fastify.listen({ port: 3000 }, function (err, address) {
    if (err) {
        fastify.log.error(err)
        process.exit(1)
    }
    log.info(`server listening on ${address}`)
})
```

**Comment by saard** [comment]
Hi,
@christiand93 Thank you for the efforts!
npm run build fails in creating schema for ConfigObject:

`projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:1077
            throw new Error("type ".concat(symbolName, " not found"));
                  ^

Error: type ConfigObject not found
    at JsonSchemaGenerator.getSchemaForSymbol (C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:1077:19)
    at generateSchema (C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:1268:26)
    at C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:1320:26 
    at step (C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:44:23)
    at Object.next (C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:25:53)
    at C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:19:71   
    at new Promise (<anonymous>)
    at __awaiter (C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:15:12)
    at exec (C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema.js:1300:12)
    at Object.run (C:\Users\I060224\projects\fastify\cf-nodejs-logging-support\node_modules\typescript-json-schema\dist\typescript-json-schema-cli.js:55:39)

Node.js v20.0.0`

**Comment by christiand93** [comment]
Hi @saard,

unfortunately, I cannot reproduce this issue on my machine (mac os sonoma) and the [build pipeline](https://github.com/SAP/cf-nodejs-logging-support/actions/runs/7446053166/job/20255543698) succeeds as well. 
I have just published a beta version ([7.3.0-beta.0](https://www.npmjs.com/package/cf-nodejs-logging-support/v/7.3.0-beta.0)) on npm, which includes the new middleware. Feel free to use it for testing.

**Comment by saard** [comment]
Hi @christiand93 
The logging works as expected for info, warn, error but - silly, debug, verbose are not logged.
I found this documentation that shows that the interface of the logger needs to be compatible with the pino interface:
https://fastify.dev/docs/v3.29.x/Reference/Server/#logger

also another issue I have found is no matter which route you are trying to reach you receive double network log:
![image](https://github.com/SAP/cf-nodejs-logging-support/assets/3519995/b7c1884c-6c05-45fe-a170-b40e29e02dc0)

Thank you!
Saar

**Comment by christiand93** [comment]
Hi @saard,
thank you for the feedback. 

> The logging works as expected for info, warn, error but - silly, debug, verbose are not logged

This sounds like the logging level threshold is set to `info`. Silly, debug, verbose logs are suppressed by default. You can change the logging level if needed:
```js
log.setLoggingLevel("silly");
```

> I found this documentation that shows that the interface of the logger needs to be compatible with the pino interface:
> https://fastify.dev/docs/v3.29.x/Reference/Server/#logger

Thank you for the pointer, I will check it out.

> also another issue I have found is no matter which route you are trying to reach you receive double network log

Are the duplicates actually logged by the cf-nodejs-logging-library? Maybe you can check the raw stdout stream to verify.
The duplicates might also be router logs. 
Logs written by our library can be identified by the `logger` field, which is always set to `nodejs-logger`.

Kind regards,
Christian

**Comment by saard** [comment]
Hi @christiand93,

Indeed the logger works and the issue was that I didn't change the default log level, sorry for that.
regarding the duplicate logging of a request, actually the nodejs-logger is printing the wrong path, and the logger named "-" is printing the correct path:
![image](https://github.com/SAP/cf-nodejs-logging-support/assets/3519995/e42c395f-1d75-4c88-a215-4ba7491b6b66)

the path that is named /healthcheck is logged by a logger called "-" (I guess no name was passed" and the "nodejs-logger" is logging the path "-" which I guess again no path was passed".

Thanks,
Saar

**Comment by christiand93** [comment]
Hi @saard,
The problem was caused by a missing configuration. I fixed the issue and also fixed another issue that was causing some response related fields to not appear. 
A new beta version is available via npm --> [7.3.0-beta.1](https://www.npmjs.com/package/cf-nodejs-logging-support/v/7.3.0-beta.1).
Next I'll do some more checking and also work on the test setup for Fastify.

Kind regards,
Christian

**Comment by saard** [comment]
@christiand93
Thank you!
Looks much better now, however there is still the duplication of log and one of them without the information:
![image](https://github.com/SAP/cf-nodejs-logging-support/assets/3519995/ae8f47b3-cc05-475e-931d-326427c10079)

![image](https://github.com/SAP/cf-nodejs-logging-support/assets/3519995/9488251a-8d02-4972-af15-372f1205b853)

Thanks,
Saar

**Comment by christiand93** [comment]
Hi @saard,
Please excuse my delayed reply. The logs in your screenshot are no duplicates, but originate from different layers. The first one got emitted by the router (RTR layer) and the second one got emitted by cf-nodejs-logging-support (app layer). You can check the `source_type` field to see which layer a log comes from.

I noticed some field type mismatches and missing fields. The problems are solved in the latest beta release ([v7.3.0-beta.2](https://www.npmjs.com/package/cf-nodejs-logging-support/v/7.3.0-beta.2)).

Kind regards,
Christian

**Comment by saard** [comment]
Hi @christiand93,

Everything looks great now 👌 
I don't recall in express the 2 logs, but I don't think it's too much of a deal.

Thank you for this great PR and hard work 💯

**Comment by christiand93** [comment]
Hi @saard,
thank you for testing and providing feedback. I will finalize the PR (tests and docs are still missing) and get the feature shipped as part of the upcoming 7.3.0 release.
